### PR TITLE
Add and update `///` comments to help consumers of these APIs

### LINKED ISSUE #47050: UseEnvironment overwritten by ConfigureWebHostDefaults
### Describe the bug
UseEnvironment is not picked up in ASP.NET Core 3.1.

I'm using `UseEnvironment` to toggle between configurations when running in Visual Studio (between on-premise and cloud versions of my service).

It worked great in v2.2, but I've recently migrated to 3.1 and it's not being used anymore.

### To Reproduce

1. Set ASPNETCORE_Environment as an environment variable 
2. Try to override it using `UseEnvironment`:

```csharp
Host.CreateDefaultBuilder(args)
    .UseEnvironment(EnvironmentName)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        webBuilder.UseStartup<Startup>();
    })
    .Build();
```

From what I understand, it cannot be set later as it won't be picked up in the Startup class then.

### Workaround

Currently I'm doing the following instead:

```csharp
Host.CreateDefaultBuilder(args)
    .ConfigureWebHostDefaults(webBuilder =>
    {
        // This overrides the default configuration
        webBuilder.UseSetting(WebHostDefaults.EnvironmentKey, EnvironmentName);
        webBuilder.UseStartup<Startup>();
    })
    .Build();
```

imho using the explicit `UseEnvironment` method should always override the default implementation (ASPNETCORE_Environment) as it's not part of the default configuration, and thus means that the dev really wants to override the default config.

I know that this has been discussed in https://github.com/dotnet/aspnetcore/issues/18499. Just saying "wont fix" leads to hours wasted trying to figure out why it isn't working. At least throw an exception `NotSupportedException` when the API is used until you can fix or remove it.

### PR DESCRIPTION
**Pull request by IEvangelist**
- Added some `///` API comments to help consumers have a better understanding of the order of operations
- Added unit tests to affirm by-design aspect of `UseEnvironment` and order of operations
- Fixed redundancy in `///` comment on `IHostEnvironment.EnvironmentName` - "of the of the"
- Added comment about `host` instance being disposed of - I didn't realize this was done implicitly and believe this is helpful to consumers

Related to and closes #47050

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs [hunk 1]
```diff
@@ -51,7 +51,8 @@ public static void Run(this IHost host)
         }
 
         /// <summary>
-        /// Runs an application and returns a Task that only completes when the token is triggered or shutdown is triggered.
+        /// Runs an application and returns a <see cref="Task"/> that only completes when the token is triggered or shutdown is triggered.
+        /// The <paramref name="host"/> instance is disposed of after running.
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs:55]: Would it make sense to go update `ConfigureWebHostDefaults` API docs? (either instead or as well)

https://github.com/dotnet/aspnetcore/blob/52eff90fbcfca39b7eb58baad597df6a99a542b0/src/DefaultBuilder/src/GenericHostBuilderExtensions.cs#L15-L30

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs [hunk 2]
```diff
@@ -141,6 +141,48 @@ public void CreateDefaultBuilder_EnablesValidateOnBuild()
...
+                })
+                .Build();
+
+            var hostEnv = host.Services.GetRequiredService<IHostEnvironment>();
+            Assert.Equal(expectedEnvironment, hostEnv.EnvironmentName);
+        }
+
+        [Theory]
+        [InlineData("Beta-Testing"), InlineData("Another-Random-Env")]
+        public void LastCallToUseEnvironmentWins(string environment)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:167]: Would it make sense to add another explicit call to `.UseEnvironment` as well?

**Review by eerhardt — APPROVED** [review]
Just a few minor thoughts to consider. Thanks for the contribution here.

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs (hunk 2 above)
Comment (IEvangelist) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:167]: There is already one call to it. Did you mean to include another, and have two calls?

```diff
                .ConfigureHostConfiguration(configBuilder =>
                {
                    configBuilder.AddInMemoryCollection(new[]
                    {
                        new KeyValuePair<string, string>(
                            HostDefaults.EnvironmentKey, willBeOverwritten)
                    });
                })
+               .UseEnvironment(someOtherEnvironment)
                .UseEnvironment(environment)
                .Build();
```

File: src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs (hunk 1 above)
Comment (IEvangelist) [reply src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs:55]: See https://github.com/dotnet/aspnetcore/pull/32025

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:167]: > have two calls

Correct. The name of the test is `LastCallToUseEnvironmentWins`. But the test only calls `UseEnvironment` once. There isn't a "previous" call to it in order to test if the "last call" wins or not.

Comment (IEvangelist) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/HostTests.cs:167]: But there is, the:

```csharp
.ConfigureHostConfiguration(configBuilder =>
{
    configBuilder.AddInMemoryCollection(new[]
    {
        new KeyValuePair<string, string>(
            HostDefaults.EnvironmentKey, willBeOverwritten)
    });
})
```

Does the same thing that `UseEnvironment` does, but I'll add another anyway.

**Pull request merged**